src/ja/common/message/base.py

```diff
@@ -3,33 +3,97 @@
 implemented in order for a class to be transferable between the components
 of JobAdder.
 """
-from abc import ABC
+from abc import ABC, abstractmethod
+from typing import Dict
 
 
 class Serializable(ABC):
     """
-    A base class for all objects which can be converted to and from a python
+    A base class for all objects which can be converted to and from a Python
     dictionary. The dictionary should contain only basic python types:
-    strings, ints, bytes, doubles, lists and dictionaries.
+    strings, ints, bytes, doubles, lists and dictionaries. The dictionary can
+    also be dumped as/read from a YAML string.
     """
 
+    @abstractmethod
+    def to_dict(self) -> Dict[str, object]:
+        """!
+        @return A dictionary representation of this object.The entries in the
+        dictionary are equivalent to the properties of this object.
+        """
 
-class Message(Serializable):
+    @classmethod
+    @abstractmethod
+    def from_dict(cls, property_dict: Dict[str, object]) -> "Serializable":
+        """!
+        @param property_dict A Python dictionary defining the desired
+        properties of the Serializable object to be created.
+        @return A new Serializable object based on the property entries of the
+        specified dictionary.
+        """
+
+    @abstractmethod
+    def to_string(self) -> str:
+        """!
+        @return A YAML document (string) representation of this object.
+        """
+
+    @classmethod
+    @abstractmethod
+    def from_string(cls, yaml_string: str) -> "Serializable":
+        """!
+        @param yaml_string A YAML document (string) defining the desired
+        properties of the Serializable object to be created.
+        @return A new Serializable object based on the properties encoded in
+        the YAML document.
+        """
+
+
+class Message(Serializable, ABC):
     """
     A base class for all messages which can be transferred between the
     components of JobAdder.
     """
 
 
-class Command(Message):
+class Command(Message, ABC):
     """
     A base class for messages which are sent with the intent of performing an
     action on the remote component.
     """
 
+    @abstractmethod
+    def get_success_response(self) -> "Response":
+        """!
+        @return A Response object indicating that the Command was successfully
+        executed.
+        """
 
-class Response(Message):
+    @abstractmethod
+    def get_failure_response(self) -> "Response":
+        """!
+        @return a Response object indicating that the Command was not
+        successfully executed.
+        """
+
+
+class Response(Message, ABC):
```

Comment:
Is the double inheritence from ABC so that we can have the appropriate symbol in the class diagram?

Reply:
It's so that Response is an abstract class.
If you inherit from an abstract class the sub-class is not automatically an abstract class, after all.